core/src/Properties/Kokkos_Detection.hpp

```diff
@@ -0,0 +1,173 @@
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+/**
+ *  Kokkos detection idiom implementation.
+ *
+ *  @todo change the macros in this file over to real detection idiom
+ *        once we drop support for compilers that don't work with it
+ */
+
+#ifndef KOKKOS_PROPERTIES_DETECTION_HPP
+#define KOKKOS_PROPERTIES_DETECTION_HPP
+
+#include <Kokkos_Core_fwd.hpp>
+
+// TODO move this to somewhere more general
+#define KOKKOS_PP_REMOVE_PARENS_IMPL(...) __VA_ARGS__
+#define KOKKOS_PP_REMOVE_PARENS(...) KOKKOS_PP_REMOVE_PARENS_IMPL __VA_ARGS__
+
+namespace Kokkos {
+namespace Impl {
+
+// A declval implementation that won't make nvcc barf
+template <class T>
+KOKKOS_FUNCTION T&& _declval(int) noexcept;
+template <class T>
+KOKKOS_FUNCTION T _declval(long) noexcept;
+
+template <class T>
+using void_t = void;
+
+template <class T>
+KOKKOS_FUNCTION decltype(Kokkos::Impl::_declval<T>(0)) declval() noexcept;
+
+struct nonesuch {
+  nonesuch()                = delete;
+  ~nonesuch()               = delete;
+  nonesuch(nonesuch const&) = delete;
+  void operator=(nonesuch const&) = delete;
+};
+
+#ifdef KOKKOS_COMPILER_SUPPORTS_DETECTION_IDIOM
+
+// primary template handles all types not supporting the archetypal Op
+template <class Default, class _always_void, template <class...> class Op,
+          class... Args>
+struct _detector {
+  constexpr static auto value = false;
+  using type                  = Default;
+};
+
+// specialization recognizes and handles only types supporting Op
+template <class Default, template <class...> class Op, class... Args>
+struct _detector<Default, void_t<Op<Args...>>, Op, Args...> {
+  constexpr static auto value = true;
+  using type                  = Op<Args...>;
+};
+
+#define KOKKOS_DECLARE_DETECTION_ARCHETYPE(name, params, params_no_intro, ...) \
+  template <KOKKOS_PP_REMOVE_PARENS(params)>
+using name = __VA_ARGS__
```

Comment:
Would that even work for the variadic argument being anything else than exactly one single argument?
Why are you using `__VA_ARGS__`?

Reply:
Are you missing a backslash??

Reply:
> Why are you using __VA_ARGS__?

In case the detection expression contains commas.  Preprocessors are weird.

> Would that even work for the variadic argument being anything else than exactly one single argument?

Yes.

> Are you missing a backslash??

Yes

Reply:
I don't get what the alias would be if there is no argument.  `void`?
And what if they are several?  Is it going to be "packed"?

Reply:
consider:

```c++
KOKKOS_DECLARE_DETECTION_ARCHETYPE(foo_archetype, (class T), (T),
  typename T::template rebind<int, char>
)
```

The preprocessor thinks the last thing is two different arguments. That's why you have to use `__VA_ARGS__` here.  There's no efficient way in the preprocessor to say "one or more", and obviously this would fail with 0.  Also, I'm missing a semicolon.

Reply:
Ban the preprocessor, resolve the conversation